google/datalab/contrib/mltoolbox/_local_predict.py

```diff
@@ -0,0 +1,167 @@
+
+  return results
+
+
+def _download_images(data, img_cols):
+  """Download images given image columns."""
+
+  images = collections.defaultdict(list)
+  for d in data:
+    for img_col in img_cols:
+      if d.get(img_col, None):
+        with file_io.FileIO(d[img_col], 'r') as fi:
+          im = Image.open(fi)
+        images[img_col].append(im)
+      else:
+        images[img_col].append('')
+
+  return images
+
+
+def _get_predicton_csv_lines(data, headers, images):
+  """Create CSV lines from list-of-dict data."""
+
+  if images:
+    data = copy.deepcopy(data)
+    for img_col in images:
+      for d, im in zip(data, images[img_col]):
+        if im == '':
+          continue
+
+        im = im.copy()
+        im.thumbnail((299, 299), Image.ANTIALIAS)
+        buf = six.StringIO()
+        im.save(buf, "JPEG")
+        content = base64.urlsafe_b64encode(buf.getvalue())
+        d[img_col] = content
+
+  csv_lines = []
+  for d in data:
+    buf = six.StringIO()
+    writer = csv.DictWriter(buf, fieldnames=headers)
+    writer.writerow(d)
+    csv_lines.append(buf.getvalue().rstrip())
+
+  return csv_lines
+
+
+def _get_display_data_with_images(data, images):
+  """Create display data by converting image urls to base64 strings."""
+
+  if not images:
+    return data
+
+  display_data = copy.deepcopy(data)
+  for img_col in images:
+    for d, im in zip(display_data, images[img_col]):
+      if im == '':
+        d[img_col + '_image'] = ''
+      else:
+        im = im.copy()
+        im.thumbnail((128, 128), Image.ANTIALIAS)
+        buf = six.StringIO()
+        im.save(buf, "PNG")
+        content = base64.b64encode(buf.getvalue())
+        d[img_col + '_image'] = content
+
+  return display_data
+
+
+def get_prediction_results(model_dir, data, headers, img_cols=None, show_image=True):
+  """ Predict with a specified model.
+
+  It predicts with the model, join source data with prediction results, and formats
+  the results so they can be displayed nicely in Datalab.
+
+  Args:
+    model_dir: The model directory.
+    data: Can be a list of dictionaries, a list of csv lines, or a Pandas DataFrame.
+    headers: the column names of data. It specifies the order of the columns when
+        serializing to csv lines.
+    img_cols: The image url columns. If specified, the img_urls will be concerted to
+        base64 encoded image bytes.
+    show_image: When displaying results, whether to add a column for showing images for
+        each image column.
+
+  Returns:
+    A dataframe of joined prediction source and prediction results.
+  """
+
+  if img_cols is None:
+    img_cols = []
+
+  if isinstance(data, pd.DataFrame):
+    data = data.T.to_dict().values()
+  elif isinstance(data[0], six.string_types):
+    data = list(csv.DictReader(data, fieldnames=headers))
+
+  images = _download_images(data, img_cols)
+  predict_data = _get_predicton_csv_lines(data, headers, images)
+  display_data = data
+  if show_image:
```

Comment:
Good point! It appears that pandas DataFrame created from a list of dict uses alphabetical order of column names, and underscore is after uppercase characters. Given prediction data: ['key', 'num', 'text', 'img_url', 'img_urlA'], it will mess up the order: ['key', 'num', 'text', 'img_url', 'img_urlA', image_urlA_image, img_url_image].

To avoid this issue, also to avoid column name clash, instead of appending "_image" now I append " (image)". Space is ordered before all non-control characters so it should work.